twine/utils.py

```diff
@@ -34,7 +34,20 @@
 DEFAULT_REPOSITORY = "https://upload.pypi.org/legacy/"
 TEST_REPOSITORY = "https://test.pypi.org/legacy/"
 
-DEFAULT_CONFIG_FILE = "~/.pypirc"
+# Obtain config file location following XDG spec
+XDG_CONFIG_HOME = os.environ.get(
+    "XDG_CONFIG_HOME", os.path.join(os.path.expanduser("~"), ".config")
+)
+XDG_CONFIG_FILE = os.path.join(XDG_CONFIG_HOME, "pypi", "pypirc")
+
+# Obtain old config file location to add backward compatibility
+OLD_CONFIG_FILE = os.path.join(os.path.expanduser("~"), ".pypirc")
+
+# Obtain config file, ${HOME}/.pypirc if exists else
+# ${XDG_CONFIG_HOME}/pypi/pypirc
+DEFAULT_CONFIG_FILE = (
+    OLD_CONFIG_FILE if os.path.isfile(OLD_CONFIG_FILE) else XDG_CONFIG_FILE
+)
```

Comment:
I'd prefer using [`pathlib`](https://docs.python.org/3/library/pathlib.html), which I think makes this more readable:

```python
XDG_CONFIG_HOME = os.environ.get("XDG_CONFIG_HOME", Path.home() /  ".config")
XDG_CONFIG_FILE = XDG_CONFIG_HOME / "pypa" / "pypirc"

HOME_CONFIG_FILE = Path.home() / ".pypirc"

DEFAULT_CONFIG_FILE = (
    HOME_CONFIG_FILE if HOME_CONFIG_FILE.is_file() else XDG_CONFIG_FILE
)